README.md
## Website README

The [200 Word RPG](https://200wordrpg.github.io/) challenge is a light game design challenge that runs every year. The goal is simple: Design a role-playing game using 200 words or less.

Every year the challenge grows larger and the entries more numerous. This post will detail some of the tools and techniques used to organize and display over 1000 entries across several years of challenges.

## Tool and Services

* [Github Pages](https://pages.github.com/) uses [Jekyll](https://jekyllrb.com/) to turn text files into web pages
* All of the text files use [Markdown](https://daringfireball.net/projects/markdown/) formatting
* Posts can be organized and sorted using [Liquid](https://github.com/Shopify/liquid/wiki) filters
* [Git](https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control) version control tracks all text and all changes
* [Python](https://www.python.org/) scripting (the script is copied at the bottom of the page)
* Random buttons inspired by [this post](https://thornelabs.net/2014/06/08/a-better-way-to-display-random-jekyll-posts-on-page-load-or-refresh-using-jquery-and-json.html) using [JSON](http://www.json.org/) database
* [Search bar](https://github.com/chinchang/super-search) stolen from [Kushagra Gour](http://kushagragour.in/)
* [Minimal theme](https://github.com/orderedlist/minimal) from [Steve Smith](https://github.com/orderedlist)
* Google [Forms](https://www.google.com/forms/about/), [Docs](https://www.google.com/docs/about/), and [Sheets](https://www.google.com/sheets/about/)

## Workflow

1. Each entry is submitted through a Google Form, which saves all submissions in a Spreadsheet.
2. I check the wordcount for each entry, making a little marker if the entry should be disqualified.
3. I download the spreadsheet as a CSV file, making sure UTF-8 encoding, and quoted text fields are enforced.
4. A python script scans the CSV file checking for duplicate entries, multiple author submissions, and the marker added.
5. The script formats all of the data and creates a Markdown file for each entry.
6. I add the markdown files to the site,and Jekyll turns it into a series of web pages.

The secret sauce is in the Python script and Jekyll.

Traditional websites like Wordpress have a SQL database where posts are stored. Whenever a user clicks a page the website looks in the database for the post info and loads it up for that user. This system is very flexible and simple to use, but it can be slow.

Jekyll is different. It doesn't have a database and any loading is all done at the beginning. Jekyll scans all of the markdown files and turns them into lean, static pages. This makes the site fast and light, but there isn't a database to query and easily alter.

The main reason I chose Jekyll, however, is because it builds pages from text files. And text files are super easy to create using python.

The python script scans through the submitted entries and turns each one into it's own Markdown text file. Through python I can check to make sure entries aren't submitted multiple times and prevent authors from submitted too many entries. Basically I can double check and filter all of the information, transforming it however I wish.

Whenever a new submission is added I can just run the python script again and have it generate a new text file for that entry. It's super quick and easy; I think it takes about 3 minutes to process ALL the entries from 2015, 2016, and 2017. Very nice.

The hard part was just getting all this stuff setup. Now that it's complete and working as expected, added new entries or changing entries is trivial and quick.

Hopefully this little guide will help you understand how the site works, and maybe even help you find uses for Jekyll and python in your own projects. It's a neat little tool and an interesting alternative to traditional website/blog site structures.

## Python Script

```
import csv
import re
import io

# A script to convert a CSV file into multiple markdown text files for Jekyll sites.
# Created by David Schirduan for the 200 Word RPG Challenge

#tracks all titles, authors, and emails to make sure the same game isn't submitted twice by the same author
allTitleEmails = []
allTitles = []
allAuthors= []


with open('2017Entries.csv', encoding="utf8") as f:
  reader = csv.reader(f)
  for row in reader:

# Rows: Timestamp, Email, Name, Link, what done?, title, entry, comments, suggestions, agreement, how hear?, validWordCount

    #To track why entries are disqualified
    invalidReason = ""
    


    #"4/14/2017 3:26:21" convert to "2017-04-14 03:26:21"
    timestamp = row[0]
    pieces = timestamp.split('/')
    
    if len(pieces[0]) == 1:
      pieces[0] = '0' + pieces[0]
      
    if len(pieces[1]) == 1:
      pieces[1] = '0' + pieces[1]

    time = pieces[2].split(' ')

    date = pieces[0] + '-' + pieces[1] + '-' + time[0]

    time = time[1]

    timepieces = time.split(':')

    if int(timepieces[0]) < 10:
      time = '0' + time

    newOrder = date.split('-')

    date = newOrder[2] + '-' + newOrder[0] + '-' + newOrder[1]

    #remove quotes from the author name. breaks MD files
    author = row[2]
    author = re.sub('\"', "", author)
    author = author.rstrip()

    #remove quotes and cap first letter in title
    title = row[5]
    title = title[0].upper() + title[1:]
    title = re.sub('\"', "", title)
    title = title.rstrip()
    fileTitle = title

    #simple way to check title and author
    email = row[1].rstrip()
    email = re.sub(',', "", email) #some people put multiple emails
    email = email.lower() #some people put caps in email
    titleEmail = title+' BY '+email

    # If title has already been submitted, check if the author is also the same
    if(title in allTitles):
      if (titleEmail in allTitleEmails):
        invalidReason = "title already submitted by this email"
        
      #if the author is different, it needs a different filename
      else:
        fileTitle = title + ' BY ' + author #to make sure the filenames are different
        print("duplicate title, different author: " + titleEmail)
        allTitleEmails.append(titleEmail)

    else:
      allTitles.append(title)
      allTitleEmails.append(titleEmail)

    #I manually checked the wordcount. Check to see if I found too many words
    if (row[11] == "invalid"):
      invalidReason = "2 many words"

    entryText = row[6]
    
    link = row[3]
    if not link:
      link = 'no link'

    comments = row[7]
    if not comments:
      comments = 'Author did not add any comments.'
    
    #Strip out all non letter and numbers for the filename
    regex = re.compile('[^a-zA-Z0-9]')
    fileSafe = regex.sub('', fileTitle)
    filename = '{0}-{1}.md'.format(date[:10],fileSafe)
    
    #Entry formatting
    line0 = '---'
    line1 = 'layout: post'
    line2 = 'title: "' + title + '"'
    line3 = 'date: ' + date + ' ' + time
    line4 = 'author: "' + author + '"'
    line5 = 'link: "' + link + '"'
    line6 = 'categories: 2017 ' + 'rpg'
    line7 = '---'
    line8 = '```'
    line9 = entryText
    line10 = '```'
    line11 = '## Author Comments '
    line12 = ''
    line13 = comments

    #as long as it isn't invalid, make a file with the content
    if (invalidReason == ""):
      if (not (author in allAuthors)):
        allAuthors.append(author)
        
      with io.open(filename,'w',encoding='utf8') as file:
        file.write(line0 + '\n')
        file.write(line1 + '\n')
        file.write(line2 + '\n')
        file.write(line3 + '\n')
        file.write(line4 + '\n')
        file.write(line5 + '\n')
        file.write(line6 + '\n')
        file.write(line7 + '\n')
        file.write(line8 + '\n')
        file.write(line9 + '\n')
        file.write(line10 + '\n')
        file.write(line11 + '\n')
        file.write(line12 + '\n')
        file.write(line13 + '\n')
        file.close

    #Otherwise print out the reason for disqualify
    else:
      print("invalid entry: " + title + ' | ' + invalidReason)

```


news.md
---
layout: default
title: News
---

Testing to see if the automated news feed works!

<ul>
{% assign sorted_pages = (site.categories.news | sort: 'date') | reverse %}
  {% for page in sorted_pages %}
      <li><h2><strong><a href="{{ page.url }}">
        {{ page.title }}
      </a></strong></h2></li>
  {% endfor %}
</ul>

script/CreateReaderAssignments.py
import csv
import re
import io

# A script to convert a CSV file into multiple markdown text files.
# Created by David Schirduan for the 200 Word RPG Challenge

# tracks all titles to make sure the same game isn't submitted twice by the same author
allTitles = []
allAuthors = []
allEmails = []
filetitle = ""

numberOfReaders = 55
i = 1

while ( i <= numberOfReaders):
    fileTitle = "2000-01-01-GROUP"+ str(i) + ".md"
    with io.open(fileTitle, 'a', encoding='utf8') as file:
        file.write('---' + '\n')
        file.write('layout: post' + '\n')
        file.write('date: 2000-01-01' + '\n')
        file.write('title: GROUP' + str(i) + '\n')
        file.write('---' + '\n')
        file.close
    i = i+1


groupCounter = 2

readerEmails = ["blinks@acm.org","zalmon135@gmail.com","aleksandra.sontowska@go.art.pl","alexandrarose.nyc@gmail.com","alexbnewcombe@gmail.com","and_lag@hotmail.com","thatonegm@gmail.com","bitte_liten_kreps@ymail.com","Cameronled@gmail.com ","Ceribythesea@gmail.com ","Petitewarlock@gmail.com","cyrus.bukowsky@gmail.com","danefogdall@gmail.com","impernious@gmail.com","sonicwaffle@outlook.com","Tsbrez@gmail.com","DCTouray@live.com","Lozsrodu8@gmail.com","valleyviolet@gmail.com","eytankaplan101@gmail.com","bluelightscribe@gmail.com","gavin.nich.rook@gmail.com","graydetrick@gmail.com","iainthecopywriter@gmail.com","jane2412@hotmail.co.uk","Jesch13@gmail.com","labyrinthalone@gmail.com","johnnie@videostorecowboy.com","mrjonathanjanssen@gmail.com","sabersger@gmail.com","k.struck@aol.com","laurenchristiansmith@gmail.com","enstong@gmail.com","marcoksk1@gmail.com","muchomil@gmail.com","zebramatt@hotmail.com","matthew@gravelyn.com","teachasneetch@juno.com","ogrebeef@gmail.com","megancondis@gmail.com","inkwan4@yahoo.com","mdlake@well.com","natalie@natalieash.com","falzon.noe@gmail.com","shimek58@gmail.com","pmjgross@gmail.com","raylvisser@gmail.com","themysteriouslever@gmail.com","dancesthroughlife@gmail.com","sean@bookseansmith.co.uk ","seth.pier@gmail.com","tara.r.king@gmail.com","thomasflott@outlook.com ","dmwt21@yahoo.com","whithughcarter@gmail.com"]

#If an entry was submitted by a Reader, it's added to Group 1, and group 1 is assigned to three readers who DIDN'T submit any entries.

with open('2019.csv', encoding="utf8") as f:
    reader = csv.reader(f)
    for row in reader:

        # Rows: Timestamp, Email, Name, Link, what done?, hear about, title, entry, comments, suggestions, agreement

        # To track why entries are disqualified
        invalidReason = ""

        timestamp = row[0]
        email = row[1].rstrip()
        author = row[2]
        pronouns = row[3]
        link = row[4]

        title = row[8]
        entryText = row[9]
        contentWarning = row[10]
        comments = row[13]

        # "4/14/2017 3:26:21" convert to "2017-04-14 03:26:21"
        pieces = timestamp.split('/')

        if len(pieces[0]) == 1:
            pieces[0] = '0' + pieces[0]

        if len(pieces[1]) == 1:
            pieces[1] = '0' + pieces[1]

        time = pieces[2].split(' ')

        date = pieces[0] + '-' + pieces[1] + '-' + time[0]

        time = time[1]

        timepieces = time.split(':')

        if int(timepieces[0]) < 10:
            time = '0' + time

        newOrder = date.split('-')

        date = newOrder[2] + '-' + newOrder[0] + '-' + newOrder[1]

        # remove quotes from the author name. breaks MD files
        author = re.sub('\"', "", author)
        author = author.rstrip()

        # The "fileTitle" is the cleaned up title for filename.
        # Need to clean up the title for sorting on the site.
        title = title[0].upper() + title[1:]
        title = re.sub('\"', '', title)
        title = title.rstrip()
        fileTitle = title

        # simple way to check title and author
        email = re.sub(',', "", email)  # some people put multiple emails
        email = email.lower()  # some people put caps in email

        # Check for existing titles, emails, and authors
        if(fileTitle in allTitles):
            if(email in allEmails or author in allAuthors):
                invalidReason = invalidReason + " | Title exists by same author/email"
            else:
                fileTitle = fileTitle + author
                allTitles.append(fileTitle)
                print(">>> VALID: " + fileTitle +
                      " Different Author, same title. ")

        else:
            allTitles.append(fileTitle)

        if(email in allEmails):
            invalidReason = invalidReason + " | Email already submitted an entry"
        else:
            allEmails.append(email)

        if(author in allAuthors):
            invalidReason = invalidReason + " | Author already submitted an entry"
        else:
            allAuthors.append(author)

        if not link:
            link = 'no link'

        if not comments:
            comments = 'Author did not add any comments.'

        # Auto-replace stupid M$ characters
        regexMS = re.compile(
            u'[\u2018\u2019\u201A\u201C\u201D\u201E\u2026\u2013\u2014\u02C6\u2039\u203A\u02DC\u00A0]+')
        results = re.findall(regexMS, title)
        if (len(results) > 0):
            title = title.replace(u'\u2018', '\'')
            title = title.replace(u'\u2019', '\'')
            title = title.replace(u'\u201A', '\'')
            title = title.replace(u'\u201C', '\"')
            title = title.replace(u'\u201D', '\"')
            title = title.replace(u'\u201E', '\"')
            title = title.replace(u'\u2026', '...')
            title = title.replace(u'\u2013', '-')
            title = title.replace(u'\u2014', '-')
            title = title.replace(u'\u02C6', '^')
            title = title.replace(u'\u2039', '<')
            title = title.replace(u'\u203A', '>')
            title = title.replace(u'\u02DC', ' ')
            title = title.replace(u'\u00A0', ' ')

        results = re.findall(regexMS, entryText)
        if (len(results) > 0):
            entryText = entryText.replace(u'\u2018', '\'')
            entryText = entryText.replace(u'\u2019', '\'')
            entryText = entryText.replace(u'\u201A', '\'')
            entryText = entryText.replace(u'\u201C', '\"')
            entryText = entryText.replace(u'\u201D', '\"')
            entryText = entryText.replace(u'\u201E', '\"')
            entryText = entryText.replace(u'\u2026', '...')
            entryText = entryText.replace(u'\u2013', '-')
            entryText = entryText.replace(u'\u2014', '-')
            entryText = entryText.replace(u'\u02C6', '^')
            entryText = entryText.replace(u'\u2039', '<')
            entryText = entryText.replace(u'\u203A', '>')
            entryText = entryText.replace(u'\u02DC', ' ')
            entryText = entryText.replace(u'\u00A0', ' ')

        # Check for invalid characters
        regexEmoji = re.compile('[^\x00-\xFF]+')

        results = re.findall(regexEmoji, title)
        chars = []
        if (len(results) > 0):
            for res in results:
                for r in list(res):
                    chars.append(ord(r))
            invalidReason = invalidReason + \
                " | Invalid Characters detected in title: " + str(chars)
            title = timestamp

        results = re.findall(regexEmoji, entryText)
        chars = []
        if (len(results) > 0):
            for res in results:
                for r in list(res):
                    chars.append(ord(r))
            invalidReason = invalidReason + \
                " | Invalid Characters detected in entry " + str(chars)

        # Wordcount the entry text
        regexCO = re.compile(
            '[0-9]+[0-9,\.]*|[&0-9a-zA-Z\xC0-\xFF]+[-\']?[0-9a-zA-Z\xC0-\xFF]*')
        results = re.findall(regexCO, entryText)
        if (len(results) > 200):
            invalidReason = invalidReason + \
                " | Wordcount: " + str(len(results))


        # Content Warning
        if (contentWarning):
            contentWarning = "<div id=\"warning\"><div id=\"content\"><h3><strong>! Content Warning: "+contentWarning + " !</strong></h3><i>Continue scrolling to read the entry.</i></div></div>"
        else:
            contentWarning = ""

        # Entry formatting
        line0 = '---'
        line1 = 'layout: post'
        line2 = 'title: "' + title + '"'
        line3 = 'date: ' + date + ' ' + time
        line4 = 'author: "' + author + '"'
        line5 = 'link: "' + link + '"'
        line6 = 'categories: 2019 rpg'
        line7 = '---'
        line8 = contentWarning
        line9 = ' '
        line10 = '```'
        line11 = entryText
        line12 = '```'
        line13 = '## Author Comments'
        line14 = comments

        # Create filename

        regex = re.compile('[^a-zA-Z0-9]')
        fileTitle = regex.sub('', fileTitle)
        fileTitle = '{0}-{1}.md'.format(date, fileTitle)

        # as long as it isn't invalid, make a file with the content
        if (invalidReason == ""):
            #If an entry was submitted by a Reader, it's added to Group 1, and group 1 is assigned to three readers who DIDN'T submit any entries.
            if (email in readerEmails):
                fileTitle = "2000-01-01-GROUP" + str(1) + ".md"
            else:
                fileTitle = "2000-01-01-GROUP" + str(groupCounter) + ".md"
                groupCounter = groupCounter + 1
                if (groupCounter > numberOfReaders):
                    groupCounter = 2
            with io.open(fileTitle, 'a', encoding='utf8') as file:
                file.write('\n# ' + title + '\n')
                file.write(line8 + '\n')
                file.write(line9 + '\n')
                file.write(line10 + '\n')
                file.write(line11 + '\n')
                file.write(line12 + '\n')
                file.write('<hr>' + '\n')
                file.close

        # Otherwise print out the reason for disqualify
        else:
            print(">>> invalid entry: " + title + invalidReason)
